bridge/setu/processor/base.go

```diff
@@ -96,3 +103,123 @@ func (bp *BaseProcessor) String() string {
 func (bp *BaseProcessor) Stop() {
 	// override to stop any go-routines in individual processors
 }
+
+// isOldTx checks if the transaction already exists in the chain or not
+// It is a generic function, which is consumed in all processors
+func (bp *BaseProcessor) isOldTx(cliCtx cliContext.CLIContext, txHash string, logIndex uint64) (bool, error) {
+	queryParam := map[string]interface{}{
+		"txhash":   txHash,
+		"logindex": logIndex,
+	}
+
+	endpoint := helper.GetHeimdallServerEndpoint(util.StakingTxStatusURL)
+	url, err := util.CreateURLWithQuery(endpoint, queryParam)
+	if err != nil {
+		bp.Logger.Error("Error in creating url", "endpoint", endpoint, "error", err)
+		return false, err
+	}
+
+	res, err := helper.FetchFromAPI(bp.cliCtx, url)
+	if err != nil {
+		bp.Logger.Error("Error fetching tx status", "url", url, "error", err)
+		return false, err
+	}
+
+	var status bool
+	if err := json.Unmarshal(res.Result, &status); err != nil {
+		bp.Logger.Error("Error unmarshalling tx status received from Heimdall Server", "error", err)
+		return false, err
+	}
+
+	return status, nil
+}
+
+// checkTxAgainstMempool checks if the transaction is already in the mempool or not
+// It is consumed only for `clerk` processor
+func (bp *BaseProcessor) checkTxAgainstMempool(msg types.Msg) (bool, error) {
+	endpoint := helper.GetConfig().TendermintRPCUrl + util.TendermintUnconfirmedTxsURL
+	resp, err := http.Get(endpoint)
+	if err != nil || resp.StatusCode != http.StatusOK {
+		bp.Logger.Error("Error fetching mempool tx", "url", endpoint, "error", err)
+		return false, err
+	}
+	body, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		bp.Logger.Error("Error fetching mempool tx", "error", err)
+		return false, err
+	}
+
+	// a minimal response of the unconfirmed txs
+	var Response struct {
+		Result struct {
+			Total string   `json:"total"`
+			Txs   []string `json:"txs"`
+		} `json:"result"`
+	}
+
+	err = json.Unmarshal(body, &Response)
+	if err != nil {
+		bp.Logger.Error("Error unmarshalling response received from Heimdall Server", "error", err)
+		return false, err
+	}
+
+	// Iterate over txs present in the mempool
+	// We can verify if the message we're about to send is present by
+	// checking the type of transaction, the transaction hash and log index
+	// present in the data of transaction
+
+	status := false
+Loop:
+	for _, txn := range Response.Result.Txs {
+		// Tendermint encodes the transactions with base64 encoding. Decode it first.
+		txBytes, err := base64.StdEncoding.DecodeString(txn)
+		if err != nil {
+			continue
```

Comment:
My bad. Will update it. Also, was wondering if we should use `Error` log, as it might appear a lot. I think it can be replaced by a warning or debug log. CC @0xKrishna.

Reply:
I don't think it will appear a lot, Use error for this./
In normal we should be able to decode a mempool txn.

Reply:
Yes, by appearing a lot, I mean if something breaks, it will appear a lot all of a sudden. Also, this is an additional check which we're doing before broadcasting a transaction, so even if this fails (due to some un-certain error), it actually doesn't cause any harm to the process.

Reply:
Have made required changes. Resolving this.